src/charm.py

```diff
@@ -4,106 +4,153 @@
+        # Defined as a property instead of an attribute in __init__ since this class is
+        # not re-instantiated between events (if there are deferred events)
+        container = self.unit.get_container(workload.Workload.CONTAINER_NAME)
+        if self.database_requires.relation:
+            return workload.AuthenticatedWorkload(
+                _container=container,
+                _admin_username=self.database_requires.relation.username,
+                _admin_password=self.database_requires.relation.password,
+                _host=self.database_requires.relation.host,
+                _port=self.database_requires.relation.port,
+                _charm=self,
+            )
+        return workload.Workload(_container=container)
 
     @property
-    def endpoint(self):
-        """The k8s endpoint for the charm."""
+    def _endpoint(self) -> str:
+        """K8s endpoint for MySQL Router"""
         return f"{self.model.app.name}.{self.model.name}.svc.cluster.local"
 
-    @property
-    def unit_hostname(self) -> str:
-        """Get the hostname.localdomain for a unit.
+    def _determine_status(self) -> ops.StatusBase:
+        """Report charm status."""
+        missing_relations = []
+        for relation in [self.database_requires, self.database_provides]:
+            if relation.missing_relation:
+                missing_relations.append(relation.NAME)
+        if missing_relations:
+            return ops.BlockedStatus(
+                f"Missing relation{'s' if len(missing_relations) > 1 else ''}: {', '.join(missing_relations)}"
+            )
+        if not self.workload.container_ready:
+            return ops.MaintenanceStatus("Waiting for container")
+        return ops.ActiveStatus()
 
-        Translate juju unit name to hostname.localdomain, necessary
-        for correct name resolution under k8s.
+    def _set_status(self) -> None:
+        """Set charm status.
 
-        Returns:
-            A string representing the hostname.localdomain of the unit.
+        Except if charm is in unrecognized state
         """
-        return f"{self.unit.name.replace('/', '-')}.{self.app.name}-endpoints"
+        if isinstance(
+            self.unit.status, ops.BlockedStatus
+        ) and not self.unit.status.message.startswith("Missing relation"):
+            return
+        self.unit.status = self._determine_status()
+        logger.debug(f"Set status to {self.unit.status}")
 
-    # =======================
-    #  Helpers
-    # =======================
+    def wait_until_mysql_router_ready(self) -> None:
+        """Wait until a connection to MySQL Router is possible.
 
-    def _patch_service(self, name: str, ro_port: int, rw_port: int) -> None:
-        """Patch juju created k8s service.
+        Retry every 5 seconds for up to 360 seconds.
+        """
+        logger.debug("Waiting until MySQL Router is ready")
+        self.unit.status = ops.WaitingStatus("MySQL Router starting")
```

Comment:
this statement may not be necessary: as far as i know, statuses are set when the handler exits, at which point the status is not necessarily `WaitingStatus`